NuRadioReco/modules/triggerTimeAdjuster.py

```diff
@@ -29,72 +29,137 @@ def begin(self, trigger_name=None, pre_trigger_time=55. * units.ns):
+                                trace = np.roll(trace, -1 * roll_by)
+                                cut_samples_beginning -= roll_by
+                            rel_station_time_samples = cut_samples_beginning + roll_by
+                        elif(samples_before_trigger > trigger_time_sample):
+                            roll_by = -trigger_time_sample + samples_before_trigger
+                            logger.info(
+                                "trigger time is before 'trigger offset window', the trace needs to be rolled by {} samples first".format(roll_by))
+                            trace = np.roll(trace, roll_by)
+                            trigger_time_sample -= roll_by
+                            rel_station_time_samples = -roll_by
 
-                    # shift trace to be in the correct location for cutting
-                    # logger.debug(f"cutting trace to {cut_samples_beginning}-{number_of_samples + cut_samples_beginning} samples")
-                    trace = trace[cut_samples_beginning:(number_of_samples + cut_samples_beginning)]
-                    channel.set_trace(trace, channel.get_sampling_rate())
-                    channel.set_trace_start_time(channel.get_trace_start_time() + rel_station_time_samples / channel.get_sampling_rate())
-                    # logger.debug(f"setting trace start time to {channel.get_trace_start_time() + rel_station_time_samples / channel.get_sampling_rate():.0f} = {channel.get_trace_start_time():.0f} + {rel_station_time_samples / channel.get_sampling_rate():.0f}")
+                        # shift trace to be in the correct location for cutting
+                        # logger.debug(f"cutting trace to {cut_samples_beginning}-{number_of_samples + cut_samples_beginning} samples")
+                        trace = trace[cut_samples_beginning:(number_of_samples + cut_samples_beginning)]
+                        channel.set_trace(trace, channel.get_sampling_rate())
+                        channel.set_trace_start_time(trigger_time)
+                        store_pre_trigger_time[channel_id] = pre_trigger_time
+                        # channel.set_trace_start_time(channel.get_trace_start_time() + rel_station_time_samples / channel.get_sampling_rate())
+                        # logger.debug(f"setting trace start time to {channel.get_trace_start_time() + rel_station_time_samples / channel.get_sampling_rate():.0f} = {channel.get_trace_start_time():.0f} + {rel_station_time_samples / channel.get_sampling_rate():.0f}")
+                
+                # store the used pre_trigger_times
+                trigger.set_pre_trigger_times(store_pre_trigger_time)
+            
+            else:
+                logger.debug('Trigger {} has not triggered. Channel timings will not be changed.'.format(self.__trigger_name))
+        elif mode == 'data_to_sim':
+            for trigger in station.get_triggers().values():
```

Comment:
This is a very implicit assumption. Then we need to make sure that `trigger_time_delays` are only filled if a trigger was the primary trigger. 
Can you at least implement a clear warning if the trace start times are corrected twice?

Reply:
Yes, this is a very good point. I thought rather than implementing some kind of flag that tells you whether or not a trigger adjusted readout windows or not it would be easier to just only fill `pre_trigger_times` if it has, but from a datataking point of view I'm not sure if that's the case, so I'm open to revisiting this if it makes more sense to implement it differently for e.g. mattak.

For now I've just added a warning.